### PR TITLE
fix(inference): align vllm_decode with LoRA checkpoint restoration an…

### PR DESCRIPTION
**Pull request by chiajunglien**
…d add test_gemma3_lora script

# Description

Start with a short description of what the PR does and how this is a change from
the past.

The rest of the description includes relevant details and context, examples:
- why is this change being made,
- the problem being solved and any relevant context,
- why this is a good solution,
- some information about the specific implementation,
- shortcomings of the solution and possible future improvements.

If the change fixes a bug or a Github issue, please include a link, e.g.,:
FIXES: b/123456
FIXES: #123456

You can also provide a comma-separated list. If you don't want to close a bug but
simply to reference it, use BUGS, e.g.: 
BUGS: b/123456

*Notice 1:* Once all tests pass, the "pull ready" label will automatically be assigned.
This label is used for administrative purposes. Please do not add it manually.

*Notice 2:* For external contributions, our settings currently require an approval from a MaxText maintainer to trigger CI tests.

# Tests

Please describe how you tested this change, and include any instructions and/or
commands to reproduce.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [ ] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [ ] I have necessary comments in my code, particularly in hard-to-understand areas.
- [ ] I have run end-to-end tests tests and provided workload links above if applicable.
- [ ] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/integration/vllm/maxtext_vllm_adapter/adapter.py [hunk 1]
```diff
@@ -323,4 +324,9 @@ def load_weights(self, rng_key: jax.Array) -> None:
       model = model_creation_utils.from_pretrained(
           self.maxtext_config, mesh=self.mesh, model_mode=self.model_mode, rng_key=rng_key
       )
+      if self.maxtext_config.lora.enable_lora:
+        model = lora_utils.apply_lora_to_model(model, self.mesh, self.maxtext_config)
+        if self.maxtext_config.lora.lora_restore_path:
+          mock_trainer = type("MockTrainer", (), {"model": model, "train_steps": 0})
```
Comment (RexBearIU) [inline src/maxtext/integration/vllm/maxtext_vllm_adapter/adapter.py]: why we mock model here? we should restore with above model like we did in train_sft

File: tests/end_to_end/tpu/gemma3/4b/test_gemma3_lora.sh [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    lora.lora_alpha=32.0 \
+    vllm_hf_overrides='{architectures: ["MaxTextForCausalLM"]}' \
+    hbm_utilization_vllm=0.5 \
+    prompt="Suggest some famous landmarks in London." \
+    use_chat_template=True \
+    enable_nnx=True \
+    pure_nnx_decoder=True \
+    scan_layers=True \
+    dtype=bfloat16 \
+    weight_dtype=bfloat16
```
Comment (RexBearIU) [inline tests/end_to_end/tpu/gemma3/4b/test_gemma3_lora.sh]: do we need those setting? please also verify.

File: tests/end_to_end/tpu/gemma3/4b/test_gemma3_lora.sh [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+# Step 5: Convert the checkpoint from MaxText format to Hugging Face format
+python3 -m maxtext.checkpoint_conversion.to_huggingface \
+    model_name=${MODEL_NAME} \
+    load_parameters_path=${BASE_OUTPUT_DIRECTORY}/sft/${run_id}/checkpoints/5/model_params \
+    lora.lora_restore_path=${BASE_OUTPUT_DIRECTORY}/lora/${run_id}/checkpoints/5/model_params \
+    base_output_directory=${BASE_OUTPUT_DIRECTORY}/to_huggingface/unscanned/${run_id} \
+    use_multimodal=false scan_layers=true
```
Comment (RexBearIU) [inline tests/end_to_end/tpu/gemma3/4b/test_gemma3_lora.sh]: why we need to set use_multimodal? And please run pre-commit to the code

File: src/maxtext/integration/vllm/maxtext_vllm_adapter/adapter.py (hunk 1 above)
Comment (chiajunglien) [reply src/maxtext/integration/vllm/maxtext_vllm_adapter/adapter.py]: I've refactored `restore_lora_from_path` in `lora_utils` to support polymorphism.
It now accepts either a full `Trainer` instance or a standalone `nnx.Module` directly.

File: tests/end_to_end/tpu/gemma3/4b/test_gemma3_lora.sh (hunk 2 above)
Comment (chiajunglien) [reply tests/end_to_end/tpu/gemma3/4b/test_gemma3_lora.sh]: Removed.

File: tests/end_to_end/tpu/gemma3/4b/test_gemma3_lora.sh (hunk 3 above)
Comment (chiajunglien) [reply tests/end_to_end/tpu/gemma3/4b/test_gemma3_lora.sh]: Removed.
